### ISSUE TITLE
FEATURE REQUEST: hash_function_from_oid()

### ISSUE THREAD

**Issue by wpd** [pyca/cryptography#11867]
I am curious to learn why (or perhaps "if", and I just missed it) some sort of `hash_function_from_oid()` function doesn't exist in the library.  So, I guess this is a not-too-subtle request to add it to the library.  Google's AI feature proposed one fairly quickly when I started wondering about this.

Anyway, I am writing code to sign CSR's for a specialized application, and I found myself writing this:
```
        csr.public_key().verify(
            csr.signature,
            csr.tbs_certrequest_bytes,
            ec.ECDSA(hashes.SHA256()),
        )
```
For this special application `ec.ECDSA(hashes.SHA256())` is the right thing to use here, but the purist in me wanted to find something like `utils.hash_from_oid(csr.signature_algorithm_oid)`.

Maybe nobody has ever asked for it before... but it seems like it should be a common sort of thing though, shouldn't it?

Maybe nobody has had the time to submit a PR to add this function to the `cryptography` library along with suitable test cases and such.  Or maybe nobody wants to maintain such a function in this library as new OID's define new algorithms.

I dunno.  But I figured I would ask.

--wpd

**Comment by alex** [comment]
https://cryptography.io/en/latest/x509/reference/#cryptography.x509.CertificateSigningRequest.signature_hash_algorithm exists on a CSR.

**Comment by wpd** [comment]
I think when I tried that, I didn't realize I needed to wrap it in a `ec.ECDSA()` call:
```
        csr.public_key().verify(
            csr.signature,
            csr.tbs_certrequest_bytes,
            ec.ECDSA(csr.signature_hash_algorithm),
        )
```
Is there a reason that, if the OID is `<ObjectIdentifier(oid=1.2.840.10045.4.3.2, name=ecdsa-with-SHA256)>` that `signature_hash_algorigthm` wouldn't wrap the `SHA256()` hash in `ec.ECDSA`?  Does it not generalize to other signature schemes?
Mainly, I was thinking it would be nice if I didn't need to know too much about the CSR (such as whether it contained an RSA or an EC key).
I guess... as I am thinking about this more... that doesn't make sense, since, if it contains an RSA key, then I also need to know what sort of padding was used.  Okay... you have convinced me :-)
